File: socket_examples/ip_notation_serverUDP.py
#!/usr/bin/env python3

import socket  # Socket module for network communication
import struct  # Struct module for packing and unpacking binary data

def main():
    # Setup UDP Socket
    sockfd = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # Create a UDP socket for IPv4
    server_address = ('0.0.0.0', 1025)  # Listen on all interfaces at port 1025
    try:
        sockfd.bind(server_address)  # Bind the socket so it can receive datagrams
    except OSError as e:
        print(f"Bind failed: {e}")
        return
    # Receive the 8-byte packet
    # data is the bytes object, addr is (ip, port) of the sender
    data, addr = sockfd.recvfrom(8)

    if len(data) < 8:
        print("Received incomplete packet")
        sockfd.close()
        return

    # Extract data using slicing
    # buffer[0:4] -> IP (4 bytes)
    # buffer[5:7] -> Port (2 bytes)
    ip_bytes = data[0:4]
    port_bytes = data[5:7]

    # Convert binary to human-readable format
    # '>I' = Unsigned Int (32-bit), Big-Endian
    # '>H' = Unsigned Short (16-bit), Big-Endian
    rec_ip_val = struct.unpack('>I', ip_bytes)[0]
    rec_port_val = struct.unpack('>H', port_bytes)[0]
    
    # Convert the 4 bytes directly to a dotted-decimal string and print
    ip_str = socket.inet_ntoa(ip_bytes)
    print(f"uint32 is {rec_ip_val}; IP is {ip_str}")
    print(f"uint16 is {rec_port_val}; Port is {rec_port_val}")

    sockfd.close()  # Close the UDP socket after processing the packet

if __name__ == "__main__":
    main()

File: socket_examples/ip_notation_serverUDP.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

int main() {
    int sockfd;  // Socket file descriptor for the UDP socket
    struct sockaddr_in servaddr, cliaddr;  // Server and client address structures
    char buffer[8];  // Buffer to receive the 8-byte packet
    socklen_t len = sizeof(cliaddr);  // Length of the client address structure

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);  // Create a UDP socket
    
    memset(&servaddr, 0, sizeof(servaddr));  // Zero out the server address structure
    servaddr.sin_family = AF_INET;  // IPv4 address family
    servaddr.sin_addr.s_addr = INADDR_ANY;  // Listen on all local IP interfaces
    servaddr.sin_port = htons(1025);  // Server port set to 1025, converted to network byte order

    if (bind(sockfd, (const struct sockaddr *)&servaddr, sizeof(servaddr)) < 0) {
        perror("Bind failed");  // Print an error if binding the socket fails
        return 1;
    }

    // Receive the 8-byte packet from any client
    recvfrom(sockfd, buffer, 8, 0, (struct sockaddr *)&cliaddr, &len);

    // Extract data from the received packet (still in network byte order)
    uint32_t rec_ip_net;
    uint16_t rec_port_net;
    memcpy(&rec_ip_net, buffer, 4);  // First 4 bytes store the IP address
    memcpy(&rec_port_net, buffer + 5, 2);  // Bytes 5-6 store the port number

    // Display the extracted values in human-readable form
    struct in_addr ip_addr;
    ip_addr.s_addr = rec_ip_net;  // Use the network-order IP bytes directly
    uint16_t final_port = ntohs(rec_port_net);  // Convert port back to host byte order

    printf("uint32 is %u; IP is %s\n", ntohl(rec_ip_net), inet_ntoa(ip_addr));
    printf("uint16 is %u; Port is %d\n", ntohs(rec_port_net), final_port);

    close(sockfd);  // Close the UDP socket when done
    return 0;
}
